File: sangsoo/0x06 queue/02_Queue_linkedlist.py
class Node:
    def __init__(self, value):
        self.val = value
        self.next = None

class Queue:
    def __init__(self):
        self.front = None
        self.rear = None
        self.count = 0

    def empty(self):
        return self.count == 0

    def push(self, val):
        newNode = Node(val)
        if self.empty():
            self.front = newNode
        else:
            self.rear.next = newNode
        self.rear = newNode
        self.count += 1

    def getFront(self):
        if self.empty():
            print("Error: Queue is empty")
            return None
        return self.front.val

    def pop(self):
        if self.empty():
            print("Error: Queue is empty")
            return
        self.front = self.front.next
        if not self.front:
            self.rear = None
        self.count -= 1

if __name__ == "__main__":
    queue = Queue()

    for i in range(1, 11):
        queue.push(i)

    while not queue.empty():
        print(queue.getFront(), end=" ")
        queue.pop()

    print()

File: sangsoo/0x06 queue/02_Queue_linkedlist.cpp
#include <iostream>
using namespace std;
class Node {
public:
    int val;
    Node* next;

    Node(int value) : val(value), next(nullptr) {}
};

class Queue {
private:
    Node* front;
    Node* rear;
    int count;

public:
    Queue() : front(nullptr), rear(nullptr), count(0) {}

    bool empty() const {
        return count == 0;
    }

    void push(int val) {
        Node* newNode = new Node(val);

        if (empty()) {
            front = newNode;
        } else {
            rear->next = newNode;
        }
        rear = newNode;
        count++;
    }

    int getFront() const {
        if (empty()) {
            cout << "Error: Queue is empty" << endl;
            return -1;
        }
        return front->val;
    }

    void pop() {
        if (empty()) {
            cout << "Error: Queue is empty" << endl;
            return;
        }
        front = front->next;
        count--;
    }
};

int main() {
    Queue queue;

    for (int i = 1; i <= 10; i++) {
        queue.push(i);
    }

    while (!queue.empty()) {
        cout << queue.getFront() << " ";
        queue.pop();
    }
    cout << endl;

    return 0;
}
